enclave/core/sgx/exception.c

```diff
@@ -385,20 +395,28 @@ void oe_virtual_exception_dispatcher(
         *arg_out = OE_EXCEPTION_CONTINUE_SEARCH;
         return;
     }
-    sgx_ssa_gpr_t* ssa_gpr =
-        (sgx_ssa_gpr_t*)(((uint8_t*)ssa_info.base_address) + ssa_info.frame_byte_size - OE_SGX_GPR_BYTE_SIZE);
+    uint64_t gprsgx_offset = (uint64_t)ssa_info.base_address +
+                             ssa_info.frame_byte_size - OE_SGX_GPR_BYTE_SIZE;
+    sgx_ssa_gpr_t* ssa_gpr = (sgx_ssa_gpr_t*)gprsgx_offset;
+    bool is_exit_info_valid = ssa_gpr->exit_info.as_fields.valid;
 
-    if (!ssa_gpr->exit_info.as_fields.valid)
+    if (!is_exit_info_valid)
     {
         // Consider manually provided exception information for SGX1
-        if (oe_exception_record)
+        // Only allow the simulated #PF in debug mode
+        if (is_enclave_debug_allowed() && oe_exception_record)
         {
             td->exception_address = ssa_gpr->rip;
 
             /* TODO PRP: We need to read the exception information from
              * oe_exception_record instead */
             td->exception_code = OE_EXCEPTION_PAGE_FAULT;
             td->exception_flags |= OE_EXCEPTION_FLAGS_HARDWARE;
+
+            /* We cannot know the faulting address that triggers #PF on SGX1
+             * unless the host passes in the information. */
+            td->faulting_address = 0;
```

Comment:
Why can't the host pass in the faulting_address info?

Reply:
Update to read the information passed in by the host. However, it seems the kernel does not allow the host to obtain the faulting address inside a user-mode signal handler; i.e., the address obtained from the signal handler on the host is always zero.

Reply:
The kernel does pass a `mcontext` to the signal handler. In there the general purpose register  [` REG_CR2`](https://code.woboq.org/userspace/glibc/sysdeps/unix/sysv/linux/x86/sys/ucontext.h.html#REG_CR2) contains the faulty address of the sigsegv. We need to add plumbing to pass in this address from OE's signal handler on the host side to the signal handler in the enclave.

Reply:
We have the dependency chain of:

 Release mode enclave -> openenclave.v4 -> this PR.

I think we should get this in quickly to move on to the other two issues.

Reply:
Just tested, the results is the same as `si_addr`.
I print the value of `REG_CR2` in the handler
```
    uint64_t faulting_address = (uint64_t)context->uc_mcontext.gregs[REG_CR2];
    printf("faulting address: 0x%lx\n", faulting_address);
```
and add an ocall to access address 0x8 at the host alone with the current two tests (access 0x8 inside the enclave and run `lgdt` instruction).

Here is the results
```
faulting address: 0x0
Test #PF passed
faulting address: 0x0
Test #GP passed
faulting address: 0x8
```

Reply:
We need to verify when the faulty address is protected, i.e., inside the enclave, the kernel still report the correct address in si_addr or CR2.

Reply:
Does you mean accessing the enclave memory by the host? it does not trigger a #PF.
- Read -> get dummy value (-1)
- Write -> no effects

Reply:
I mean changing the page permission on an enclave address to, for example, read only, and then writing to that address in the enclave code. Fundamentally this is the absolute dominate use case for #PF events in Mystikos.

Reply:
I can confirm that CR2/si_addr is not cleared if #PF is on an enclave memory.
I test with writing to an enclave code page.
```
enclave code addr: 0x7fc603003500
faulting address: 0x7fc603003000
```

Reply:
Let me update the comments and test cases accordingly.